src/engine/input/RunOnKeyAttribute.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+
+public class RunOnKeyAttribute : RunOnInputAttribute
+{
+    private IInputReceiver inputReceiver;
```

Comment:
It seems like this is needed by both classes inheriting RunOnInputAttribute. I think you could make this a property in the parent class, and make an abstract protected method `CreateInputReceiver` that these two classes inherit, to reduce the needed logic a bit.